trio/_tests/tools/test_gen_exports.py

```diff
@@ -119,12 +122,65 @@ def test_process(tmp_path, imports):
     assert excinfo.value.code == 1
 
 
+@skip_lints
+def test_run_black(tmp_path) -> None:
+    """Test that processing properly fails if black does."""
+    try:
+        import black  # noqa: F401
+    except ImportError as error:  # pragma: no cover
+        skip_if_optional_else_raise(error)
+
+    file = File(tmp_path / "module.py", "module")
+
+    success, _ = run_black(file, "class not valid code ><")
+    assert not success
+
+    success, _ = run_black(file, "import waffle\n;import trio")
+    assert not success
+
+
+@skip_lints
+def test_run_ruff(tmp_path) -> None:
+    """Test that processing properly fails if black does."""
+    try:
+        import ruff  # noqa: F401
+    except ImportError as error:  # pragma: no cover
+        skip_if_optional_else_raise(error)
+
+    file = File(tmp_path / "module.py", "module")
+
+    success, _ = run_ruff(file, "class not valid code ><")
+    assert not success
+
+    success, response = run_ruff(
+        file,
+        '''def combine_and(data: list[str]) -> str:
+    """Join values of text, and have 'and' with the last one properly."""
+    if len(data) >= 2:
+        data[-1] = 'and ' + data[-1]
+    if len(data) > 2:
+        return ', '.join(data)
+    return ' '.join(data)''',
+    )
+    assert success
+    assert (
+        response
+        == '''def combine_and(data: list[str]) -> str:
+    """Join values of text, and have 'and' with the last one properly."""
+    if len(data) >= 2:
+        data[-1] = 'and ' + data[-1]
+    if len(data) > 2:
+        return ', '.join(data)
+    return ' '.join(data)'''
```

Comment:
To my untrained eye this looks the same as the file above. Maybe maybe a variable so you only have to type it once?

Also, just wondering where this sample code comes from!

Reply:
Checked and it is indeed identical, will make a variable shortly.

Sample code was in one of the files I have on my desktop when I'm testing things, but digging deeper it looks like a modified version of one of the functions from my discord bot (https://github.com/CoolCat467/StatusBot/blob/4c4a0808ecbba4bcd3a1a6f0bff8169bc8a2bbd5/src/StatusBot/bot.py#L169C1-L176C26)

Reply:
Made variable in af0547ae